Method 1: support python 2 and python 3 by automatically converting parameters to 8-bit strings.

tacitus-aedifex opened the pull request:
Hi Stef,

Here's what I think is a super clean PR to add support for python 3 without interfering with python 2 operation.  I modified the encode_strings decorator to be sensitive to the python version and it correctly handles encoding unicode strings into strings of 8-bit values before passing them to libsodium.

The unit tests all pass on both python2 and python3.

//tæ

stef commented:
i'm not sure i'm in favor of handling stringencoding of two languages in one library wrapper. can't we just require the developer himself to ensure the correct types are passed? i'd rather change the testcases so they always provide the correct string, and thus keep out the string handling confusion of two languages out of the main functionality.

tacitus-aedifex commented:
If you don't want to handle the string conversion in the library then I think the correct thing to do is:

1. remove all of the calls to .encode() inside of the functions.
2. add a decorator that raises ValueError exceptions when passed unicode objects in python2 and str objects in python3.

What do you think?

tacitus-aedifex commented:
fixed a few places that I missed that need encode_strings.  i also cleaned up the test cases so that they all pass binary strings to the pysodium functions.

tacitus-aedifex commented:
@stef I've been using both versions when working on a project that relies on pysodium and I have to say this method is much more convenient and more resistant to programmer error.  this is the direction I think pysodium should take.

tacitus-aedifex commented:
@stef ping?

stef commented:
sorry man. i have lot's of respect for your work, but this is such a mess - exactly the reason why i don't support py3. i don't want to prescribe to the user what kind of bytes they put into nacl, for nacl the strings are totally opaque, nacl gives no fuck about stupid python encodings and if the coder want's to shoot themselves in the foot, we should blame the fucking language designers, not putting stupid workarounds in every damn module there is. if the coder using pysodium has no clue about what kind of bytes it feeds to nacl, then maybe the coder should not use these technologies except for shooting themselves in the foot. i mean pysodium is a wrapper not a bloated nanny. i'm not going to merge this. :/

sorry tacitus, i had to vent (especially after you pinging me). hope you understand.

tacitus-aedifex commented:
@stef, no worries, I understand and respect your opinion on the subject.  I agree that the string encoding stuff is bogus.  I need python3 support so I'm going to maintain my own fork that tracks your project.

evadot commented:
I use pysodium with python3 in production, no problem. You just need to pass correct arguments to the function.
Don't worry that I'll be the first to complain if py3 support suddenly broke :)

stef commented:
that's what i mean, let the dev decide what bytes he inputs into the sodium stuff, sodium really doesn't care if it's py2 or py3 or if it's some stupid bytes, or strings, or unicode.
